### PR TITLE
[MLDM-20] createJob RPC and tests

### PR DESCRIPTION
**Pull request by Zhang-Muyang**
first pass of createJob and InspectJob.
Todo: 1. add context
2. Test inspectJob when Job is in Processing and Done state.

### CODE REVIEW

File: src/internal/pjs/api_server.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	}
+}
+
+func (a *apiServer) CreateJob(ctx context.Context, request *pjs.CreateJobRequest) (response *pjs.CreateJobResponse, retErr error) {
+	var ret pjs.CreateJobResponse
+	if request.Input == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "missing data input")
+	}
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		program, err := fileset.ParseID(request.Program)
```
Comment (brendoncarroll) [inline src/internal/pjs/api_server.go:36]: This stuff doesn't need to be in the transaction; you can move it above so the transaction is just the operation you need to perform.

File: src/internal/pjs/api_server.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			inputs = append(inputs, fileset.PinnedFileset(*inputID))
+		}
+		req := pjsdb.CreateJobRequest{
+			Program: fileset.PinnedFileset(*program),
+			Inputs:  inputs,
+		}
+		jobID, err := pjsdb.CreateJob(ctx, sqlTx, req)
```
Comment (brendoncarroll) [inline src/internal/pjs/api_server.go:52]: This has to stay.

File: src/internal/pjs/api_server.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+			ParentJob: &pjs.Job{
+				Id: int64(job.Parent),
+			},
+			Program: job.Program.HexString(),
+			State:   pjs.JobState_QUEUED,
+		}
+		for _, filesetID := range job.Inputs {
+			jobInfo.Input = append(jobInfo.Input, filesetID.HexString())
+		}
+		if !job.Processing.IsZero() {
```
Comment (brendoncarroll) [inline src/internal/pjs/api_server.go:82]: 82-88 here can be in a switch block.   In Go it's more idiomatic  to write sequences of mutually exclusive if-elses as a switch.

```go
switch {
case ifFirstCaseIsTrue:
case ifSecondCaseIsTrue:
default:
  return fmt.Errorf("invalid job from db: %v")
```

File: src/internal/pjs/server.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import (
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage"
+	txnenv "github.com/pachyderm/pachyderm/v2/src/internal/transactionenv"
+	pjsserver "github.com/pachyderm/pachyderm/v2/src/pjs"
+)
+
+type Env struct {
+	DB      *pachsql.DB
+	TxnEnv  *txnenv.TransactionEnv
```
Comment (brendoncarroll) [inline src/internal/pjs/server.go:12]: Remove

File: src/internal/pjs/server.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	Storage *storage.Server
```
Comment (brendoncarroll) [inline src/internal/pjs/server.go:13]: Remove for now.  I'm not sure if we will have to pull this back in.  I think we should only need `fileset.Pin`, no the entire storage layer.  Either way we shouldn't have unused dependencies.

File: src/internal/pjs/server_test.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+func TestInspectJob(t *testing.T) {
+	t.Run("valid", func(t *testing.T) {
+		c, fc := setupTest(t)
+		ctx := pctx.TestContext(t)
+		programFileset := createFileSet(t, fc, map[string][]byte{
+			"file/path": []byte(`!#/bin/bash; ls /input/;`),
+		})
+		inputFileset := createFileSet(t, fc, map[string][]byte{
+			"a.txt": []byte(`dummy input`),
+		})
```
Comment (brendoncarroll) [inline src/internal/pjs/server_test.go:36]: InspectJob should error if it is called before the first create.  Can we add a case for that?

We should return an error with the not found code.
https://pkg.go.dev/google.golang.org/grpc/codes#Code

File: src/internal/storage/util.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	s, err := New(ctx, Env{
+		DB:     db,
+		Bucket: b,
+		Config: pachconfig.StorageConfiguration{},
+	})
+	require.NoError(t, err)
+	return s
+}
+
+func NewTestFilesetClient(t testing.TB, db *pachsql.DB) storage.FilesetClient {
```
Comment (brendoncarroll) [inline src/internal/storage/util.go:28]: 👍

File: src/internal/pjs/api_server.go (hunk 3 above)
Comment (Zhang-Muyang) [reply src/internal/pjs/api_server.go:82]: @brendoncarroll 
line 88 handles job result if state is done. Line 82-87 is not mutually exclusive. A job in "Done" state also has processing timestamp and queued timestamp, so line 85 and 82 can be correct at the same time.

Comment (brendoncarroll) [reply src/internal/pjs/api_server.go:82]: A job can only be in one state at a time though.  `jobInfo.State` can only be assigned to one value after all is said and done.  If we determine what that is by checking timestamps, and there are some database states that are unaccounted for, then we should be able to detect when they occur.

**Review by brendoncarroll — APPROVED** [review]
LGTM

File: src/internal/pjs/api_server.go (hunk 3 above)
Comment (Zhang-Muyang) [reply src/internal/pjs/api_server.go:82]: Got it. I'll fix this. I shouldn't assume all the states are valid.

**Pull request merged**